File: src/KeyVault/shamir_share/python/shared_secret.py
import array
from mod_math import *

max_shares = 126

class make_shared_secret:
    def __init__(self, shares, required):
        if shares > max_shares or required > shares or required < 2:
            raise Exception("Incorrect share or required count")

        self.shares = shares
        self.required = required
        self.byte_shares = make_byte_shares( required, 0)

    def make_byte_shares(self, b):
        share_array = []
        self.byte_shares.set_secret_byte(b)

        x = 1
        for i in range(self.shares):
            s = self.byte_shares.make_share(x)
            share_array.append(s.to_uint16())
            x = x + 1

        return share_array

    def make_shares(self, plaintext):
        share_arrays = []

        for i in range( len(plaintext) ):
            p = plaintext[i]
            share_array = self.make_byte_shares(p)

            share_count = len(share_array)

            for j in range(share_count):
                if i == 0:
                    tmp = array.array('H')
                    share_arrays.append(tmp)

                current_share_array = share_arrays[j]
                current_share_array.append(share_array[j])

        return share_arrays

class get_secret:
    def __init__(self, required):
        self.required = required
        self.gbs = get_byte_secret()

    def get_secret_byte(self, share_array):
        if len(share_array) < self.required:
            raise Exception("Insufficient shares")

        return self.gbs.get_secret(share_array, self.required)

    def get_plaintext(self, share_arrays):
        plaintext = bytearray()
        plaintext_len = len(share_arrays[0])

        if len(share_arrays) < self.required:
            raise Exception("Insufficient shares")

        for j in range(plaintext_len):
            sv = array.array('H')

            for i in range(self.required):
                sa = share_arrays[i]
                sv.append(sa[j])

            text = self.get_secret_byte(sv)
            plaintext.append(text)

        return plaintext



    

File: src/KeyVault/KeyVault/SecurityDomain/Crypto/shared_secret.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Microsoft.Azure.Commands.KeyVault.SecurityDomain.Crypto
{
    public class shared_secret
	{
		public shared_secret(UInt16 shares, UInt16 required)
        {
			if (shares > max_shares || required > shares || required < 2)
				throw new Exception("Incorrect share or required count");

			_shares = shares;
			_required = required;
			_coefficients = new UInt16[required];
			_rand_bits = new random_bits(64);

		}

		public shared_secret(UInt16 required)
		{
			if (required < 2)
				throw new Exception("Incorrect share or required count");

			_shares = 0;
			_required = required;
			_rand_bits = new random_bits(64);
		}

		public List<UInt16[]> make_shares(byte[] plaintext)
		{
			// Output will have size of share count, each share vector will have an entry for every input byte
			List<UInt16[]> share_arrays = new List<UInt16[]>();

			for( UInt32 i = 0; i < plaintext.Length; ++i)
			{
				byte p = plaintext[i];
				UInt16[] share_array = make_shares(p);

				/*
				We now have a share created for the total number of shares needed
				Each share then needs to be distributed such that there's a share 
				for each byte of plaintext, effectively transposing the 2-dimensional 
				array.
				*/
				Int32 share_count = share_array.Length;

				for (Int32 j = 0; j < share_count; ++j)
				{
					if (i == 0)
						share_arrays.Add(new UInt16[plaintext.Length]);

					UInt16[] current_share_array = share_arrays[j];
					current_share_array[i] = share_array[j];
				}
			}

			return share_arrays;
		}

		public UInt16[] make_shares(byte secret_byte)
		{
			UInt16[] share_array = new UInt16[_shares];

			init_coefficients();
			_coefficients[(UInt32)(_required) - 1] = secret_byte;

			UInt16 x = 1;
			for (UInt32 i = 0; i < _shares; ++i, ++x)
			{
				share s = new share(x, shared_math.make_share(_coefficients, x));
				share_array[i] = s.to_uint16();
			}

			return share_array;
		}

		public byte[] get_secret(List<UInt16[]> share_arrays)
		{
			byte[] plaintext = new byte[share_arrays[0].Length];

			if (share_arrays.Count < _required)
			{
				throw new Exception("Insufficient shares");
			}

			UInt16[] sv = new UInt16[_required];
	
			// TODO - all the constants calculated in get_secret
			// can be pulled out once and re-used, which will help perf
			for (UInt32 j = 0; j < plaintext.Length; ++j)
			{
				for (Int32 i = 0; i< _required; ++i)
				{
					UInt16[] sa = share_arrays[i];
					sv[i] =  sa[j];
				}

				UInt16 text = get_secret(sv);
				plaintext[j] = (byte)(text);
			}

			return plaintext;
		}

		public UInt16 get_secret(UInt16[] share_array)
        {
			if (share_array.Length < _required)
				throw new Exception("Insufficient shares");

			return shared_math.get_secret(share_array, _required);
		}

		void init_coefficients()
		{
			for (UInt32 i = 0; i < (UInt32)(_required) - 1; ++i)
			{
				_coefficients[i] = _rand_bits.get_mod_257();
			}
		}

		UInt16 _shares;
		UInt16 _required;
		UInt16[] _coefficients;
		random_bits _rand_bits;

		const UInt32 max_shares = 126;
	}
}
